modules/apps/27-interchain-accounts/host/keeper/relay_test.go

```diff
@@ -451,13 +454,363 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
 		},
 	}
 
+	for _, encoding := range testedEncodings {
+		for _, tc := range testCases {
+			tc := tc
+
+			suite.Run(tc.msg, func() {
+				suite.SetupTest() // reset
+
+				path = NewICAPath(suite.chainA, suite.chainB, encoding)
+				suite.coordinator.SetupConnections(path)
+
+				err := SetupICAPath(path, TestOwnerAddress)
+				suite.Require().NoError(err)
+
+				portID, err := icatypes.NewControllerPortID(TestOwnerAddress)
+				suite.Require().NoError(err)
+
+				// Get the address of the interchain account stored in state during handshake step
+				storedAddr, found := suite.chainB.GetSimApp().ICAHostKeeper.GetInterchainAccountAddress(suite.chainB.GetContext(), ibctesting.FirstConnectionID, portID)
+				suite.Require().True(found)
+
+				icaAddr, err := sdk.AccAddressFromBech32(storedAddr)
+				suite.Require().NoError(err)
+
+				// Check if account is created
+				interchainAccount := suite.chainB.GetSimApp().AccountKeeper.GetAccount(suite.chainB.GetContext(), icaAddr)
+				suite.Require().Equal(interchainAccount.GetAddress().String(), storedAddr)
+
+				suite.fundICAWallet(suite.chainB.GetContext(), path.EndpointA.ChannelConfig.PortID, sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdkmath.NewInt(10000))))
+
+				tc.malleate(encoding) // malleate mutates test data
+
+				packet := channeltypes.NewPacket(
+					packetData,
+					suite.chainA.SenderAccount.GetSequence(),
+					path.EndpointA.ChannelConfig.PortID,
+					path.EndpointA.ChannelID,
+					path.EndpointB.ChannelConfig.PortID,
+					path.EndpointB.ChannelID,
+					suite.chainB.GetTimeoutHeight(),
+					0,
+				)
+
+				txResponse, err := suite.chainB.GetSimApp().ICAHostKeeper.OnRecvPacket(suite.chainB.GetContext(), packet)
+
+				if tc.expPass {
+					suite.Require().NoError(err)
+					suite.Require().NotNil(txResponse)
+				} else {
+					suite.Require().Error(err)
+					suite.Require().Nil(txResponse)
+				}
+			})
+		}
+	}
+}
+
+func (suite *KeeperTestSuite) TestJSONOnRecvPacket() {
```

Comment:
Is this the same as `TestOnRecvPacket`, but instead of calling `SerializeCosmosTx`, the bytes are being manually set? Would it be simpler to create a test which just asserts the manually created bytes == the output of `SerializeCosmosTx`. Is this test adding code coverage in a different way?

Reply:
That's right. I think doing what you said in `codec_test.go` would be more or less equivalent. However, currently, I think it does increase codecov as I added test that weren't included in `TestOnRecvPacket`. I think having this here is still beneficial because:

 1. Redundancy in tests is good.
 2. It ensures that that the json bytes that are created outside of the sdk's codec can actually be executed.
 3. These jsons are actually not identical to ones that would be produced by sdk's codec.
 
 But still open to moving them. I can only see upside to keeping them here.